src/udev/dmi_memory_id/dmi_memory_id.c

```diff
@@ -145,123 +130,136 @@ static void dmi_print_memory_size(
+                [0x02] = "PC-98/E Add-on Card",         /* 0xA2 */
+                [0x03] = "PC-98/Local Bus Add-on Card", /* 0xA3 */
+                [0x04] = "CXL Flexbus 1.0",             /* 0xA4 */
         };
+        const char *str = OUT_OF_SPEC_STR;
+
+        if (code < ELEMENTSOF(location) && location[code])
+                str = location[code];
+        else if (code >= 0xA0 && code < (ELEMENTSOF(location_0xA0) + 0xA0))
+                str = location_0xA0[code - 0xA0];
 
-        if (code < ELEMENTSOF(location) && code[location])
-                return location[code];
-        if (code >= 0xA0 && code < (ELEMENTSOF(location_0xA0) + 0xA0))
-                return location_0xA0[code - 0xA0];
-        return OUT_OF_SPEC_STR;
+        printf("MEMORY_ARRAY_LOCATION=%s\n", str);
 }
 
-static const char *dmi_memory_array_ec_type(uint8_t code) {
+static void dmi_memory_array_ec_type(uint8_t code) {
         /* 7.17.3 */
         static const char *type[] = {
                 [0x01] = "Other",
-                "Unknown",
-                "None",
-                "Parity",
-                "Single-bit ECC",
-                "Multi-bit ECC",
-                "CRC" /* 0x07 */
+                [0x02] = "Unknown",
+                [0x03] = "None",
+                [0x04] = "Parity",
+                [0x05] = "Single-bit ECC",
+                [0x06] = "Multi-bit ECC",
+                [0x07] = "CRC",
         };
 
-        if (code < ELEMENTSOF(type) && code[type])
-                return type[code];
-        return OUT_OF_SPEC_STR;
+        if (code != 0x03) /* Do not print "None". */
+                printf("MEMORY_ARRAY_EC_TYPE=%s\n",
+                       code < ELEMENTSOF(type) && type[code] ? type[code] : OUT_OF_SPEC_STR);
 }
 
 /*
  * 7.18 Memory Device (Type 17)
  */
 
+static void dmi_memory_device_string(
+                const char *attr_suffix, unsigned slot_num,
+                const struct dmi_header *h, uint8_t s) {
+        char *str;
+
+        str = strdupa(dmi_string(h, s));
+        printf("MEMORY_DEVICE_%u_%s=%s\n", slot_num, attr_suffix, strstrip(str));
+}
+
 static void dmi_memory_device_width(
-                const char *attr_prefix, const char *attr_suffix,
+                const char *attr_suffix,
                 unsigned slot_num, uint16_t code) {
-        /*
-         * If no memory module is present, width may be 0
-         */
-        if (code != 0xFFFF && code != 0)
-                printf("%s_%u_%s=%u\n", attr_prefix, slot_num, attr_suffix, code);
+
+        /* If no memory module is present, width may be 0 */
+        if (!IN_SET(code, 0, 0xFFFF))
+                printf("MEMORY_DEVICE_%u_%s=%u\n", slot_num, attr_suffix, code);
 }
 
 static void dmi_memory_device_size(unsigned slot_num, uint16_t code) {
-        if (code == 0) {
-                printf("MEMORY_DEVICE_%u_PRESENT=0\n", slot_num);
-                return;
-        } else if (code == 0xFFFF) {
+        uint64_t s = code & 0x7FFF;
```

Comment:
I would only have initialised it before being used.

Reply:
OK.